ql/math/matrixutilities/gmres.cpp

```diff
@@ -0,0 +1,155 @@
+*/
+
+/*! \file gmres.cpp
+    \brief generalized minimal residual method
+*/
+
+
+#include <ql/math/functional.hpp>
+#include <ql/math/matrixutilities/gmres.hpp>
+#include <ql/math/matrixutilities/qrdecomposition.hpp>
+
+#include <numeric>
+
+namespace QuantLib {
+
+    GMRES::GMRES(const GMRES::MatrixMult& A,
+                 Size maxIter, Real relTol,
+                 const GMRES::MatrixMult& preConditioner)
+    : A_(A), M_(preConditioner),
+      maxIter_(maxIter), relTol_(relTol) {
+
+        QL_REQUIRE(maxIter_ > 0, "maxIter must be greater then zero");
+    }
+
+    GMRESResult GMRES::solve(const Array& b, const Array& x0) const {
+        const GMRESResult result = solveImpl(b, x0);
+
+        QL_REQUIRE(result.errors.back() < relTol_, "could not converge");
+
+        return result;
+    }
+
+    GMRESResult GMRES::solveWithRestart(
+        Size restart, const Array& b, const Array& x0) const {
+
+        GMRESResult result = solveImpl(b, x0);
+
+        std::list<Real> errors = result.errors;
+
+        for (Size i=0; i < restart-1 && result.errors.back() >= relTol_;++i) {
+            result = solveImpl(b, result.x);
+            errors.insert(
+                errors.end(), result.errors.begin(), result.errors.end());
+        }
+
+        QL_REQUIRE(errors.back() < relTol_, "could not converge");
+
+        result.errors = errors;
+        return result;
+    }
+
+    GMRESResult GMRES::solveImpl(const Array& b, const Array& x0) const {
+        const Real bn = norm2(b);
+        if (bn == 0.0) {
+            GMRESResult result = { std::list<Real>(1, 0.0), b };
+            return result;
+        }
+
+        Array x = ((!x0.empty()) ? x0 : Array(b.size(), 0.0));
+        Array r = b - A_(x);
+
+        const Real g = norm2(r);
+        if (g/bn < relTol_) {
+            GMRESResult result = { std::list<Real>(1, g/bn), x };
+            return result;
+        }
+
+        std::vector<Array> v(1, r/g);
+        std::vector<Array> h(1, Array(maxIter_, 0.0));
+        std::vector<Real>  c(maxIter_+1), s(maxIter_+1), z(maxIter_+1);
+
+        z[0] = g;
+
+        std::list<Real> errors(1, g/bn);
+
+        for (Size j=0; j < maxIter_ && errors.back() >= relTol_; ++j) {
+            h.push_back(Array(maxIter_, 0.0));
+            Array w = A_((M_)? M_(v[j]) : v[j]);
+
+            for (Size i=0; i <= j; ++i) {
+                h[i][j] = DotProduct(w, v[i]);
+                w -= h[i][j] * v[i];
+            }
+
+            h[j+1][j] = norm2(w);
+
+            if (h[j+1][j] < QL_EPSILON*QL_EPSILON)
+                break;
+
+            v.push_back(w / h[j+1][j]);
+
+            for (Size i=0; i < j; ++i) {
+                const Real h0 = c[i]*h[i][j] + s[i]*h[i+1][j];
+                const Real h1 =-s[i]*h[i][j] + c[i]*h[i+1][j];
+
+                h[i][j]   = h0;
+                h[i+1][j] = h1;
+            }
+
+            const Real tau = std::fabs(h[j][j]) + std::fabs(h[j+1][j]);
```

Comment:
This is reported as unused.  Should we remove it, or is it missing somewhere later?